cloudapi/insights/client.go

```diff
@@ -0,0 +1,206 @@
+package insights
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"go.k6.io/k6/cloudapi/insights/proto/v1/ingester"
+)
+
+const (
+	testRunIDHeader     = "X-K6TestRun-Id"
+	authorizationHeader = "Authorization"
+)
+
+var (
+	// ErrClientAlreadyInitialized is returned when the client is already initialized.
+	ErrClientAlreadyInitialized = errors.New("insights client already initialized")
+
+	// ErrClientClosed is returned when the client is closed.
+	ErrClientClosed = errors.New("insights client closed")
+)
+
+// ClientConfig is the configuration for the client.
+type ClientConfig struct {
+	IngesterHost  string
+	ConnectConfig ClientConnectConfig
+	AuthConfig    ClientAuthConfig
+	TLSConfig     ClientTLSConfig
+}
+
+// ClientConnectConfig is the configuration for the client connection.
+type ClientConnectConfig struct {
+	Block                  bool
+	FailOnNonTempDialError bool
+	Dialer                 func(context.Context, string) (net.Conn, error)
+}
+
+// ClientAuthConfig is the configuration for the client authentication.
+type ClientAuthConfig struct {
+	Enabled                  bool
+	TestRunID                int64
+	Token                    string
+	RequireTransportSecurity bool
+}
+
+// ClientTLSConfig is the configuration for the client TLS.
+type ClientTLSConfig struct {
+	Insecure bool
+	CertFile string
+}
+
+// Client is the client for the k6 Insights ingester service.
+type Client struct {
+	cfg    ClientConfig
+	client ingester.IngesterServiceClient
+	conn   *grpc.ClientConn
+	connMu *sync.RWMutex
+}
+
+// NewClient creates a new client.
+func NewClient(cfg ClientConfig) *Client {
+	return &Client{
+		cfg:    cfg,
+		client: nil,
+		conn:   nil,
+		connMu: &sync.RWMutex{},
+	}
+}
+
+// Dial creates a client connection using ClientConfig.
+func (c *Client) Dial(ctx context.Context) error {
+	c.connMu.Lock()
+	defer c.connMu.Unlock()
+
+	if c.conn != nil {
+		return ErrClientAlreadyInitialized
+	}
+
+	opts, err := dialOptionsFromClientConfig(c.cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create dial options: %w", err)
+	}
+
+	conn, err := grpc.DialContext(ctx, c.cfg.IngesterHost, opts...)
+	if err != nil {
+		return fmt.Errorf("failed to dial: %w", err)
+	}
+
+	c.client = ingester.NewIngesterServiceClient(conn)
+	c.conn = conn
+
+	return nil
+}
+
+// IngestRequestMetadatasBatch ingests a batch of request metadatas.
+func (c *Client) IngestRequestMetadatasBatch(ctx context.Context, requestMetadatas RequestMetadatas) error {
+	c.connMu.RLock()
+	if c.conn == nil {
+		return ErrClientClosed
+	}
+	c.connMu.RUnlock()
+
+	if len(requestMetadatas) < 1 {
+		return nil
+	}
+
+	req, err := newBatchCreateRequestMetadatasRequest(requestMetadatas)
+	if err != nil {
+		return fmt.Errorf("failed to create request from request metadatas: %w", err)
```

Comment:
```suggestion
		return fmt.Errorf("failed to create request from request metadata: %w", err)
```
Metadata doesn't have the plural, am I missing something? I see it is used all over the code so if we change then keep it consistent.

Reply:
Yes, we are aware it doesn't have a plural form. However, together with @vortegatorres we decided to add the `s` suffix to match the conventions defined in [Google API Design Guide](https://cloud.google.com/apis/design). I'd like to keep this as "request metadatas".